FakeStorageRpc - can't handle large amounts of data due to regex error for INCOMPLETE contentRangeHeader pattern

**Steps to reproduce**
1. Create a UnitTest that uses the com.google.cloud.google-cloud-nio FakeStorageRpc
2. Create a storage
3. Create an InputStream which streams a large amount of data
4. Create a Blob handling the InputStream

**Code example**
Here is a [reproducer](https://github.com/KaminskiDaniel0/FakeStorageRpcContentRangeIssue) - s. `ApplicationTests.testCreateLargeBlob()`

Basically the test looks like this:
```java
@Test
void testCreateLargeBlob() throws Exception {

        try (final InputStream inPipe = createLargeDataStream(TOTAL_FILE_SIZE)) {

            final Blob blob = storage.createFrom(
                BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, FILE_NAME)).build(),
                inPipe,
                BUFFER_SIZE
            );

            assertNotNull(blob);
            assertEquals(BUCKET_NAME, blob.getBucket());
            assertEquals(FILE_NAME, blob.getName());
        }
    }
```


**Stacktrace**
```
java.io.IOException: com.google.cloud.storage.StorageException: 404 
	|> PUT http://localhost:65555/b/my-staging-bucket/o/fake_file_name.zip
	|> content-range: bytes 0-262143/*
	|  
	|< HTTP/1.1 404
	|  

	at com.google.cloud.storage.BaseStorageWriteChannel.write(BaseStorageWriteChannel.java:122)
	at com.google.cloud.storage.StorageImpl.uploadHelper(StorageImpl.java:325)
	at com.google.cloud.storage.StorageImpl.createFrom(StorageImpl.java:300)
	at com.dkaminski.fakestoragerpccontentrangeissue.ApplicationTests.testCreateLargeBlob(ApplicationTests.java:40)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
Caused by: com.google.cloud.storage.StorageException: 404 
	|> PUT http://localhost:65555/b/my-staging-bucket/o/fake_file_name.zip
	|> content-range: bytes 0-262143/*
	|  
	|< HTTP/1.1 404
	|  
	at com.google.cloud.storage.ResumableSessionFailureScenario.toStorageException(ResumableSessionFailureScenario.java:280)
	at com.google.cloud.storage.ResumableSessionFailureScenario.toStorageException(ResumableSessionFailureScenario.java:146)
	at com.google.cloud.storage.JsonResumableSessionPutTask.call(JsonResumableSessionPutTask.java:218)
	at com.google.cloud.storage.JsonResumableSession.lambda$put$0(JsonResumableSession.java:81)
	at com.google.cloud.storage.Retrying.lambda$run$0(Retrying.java:102)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:102)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.Retrying.run(Retrying.java:99)
	at com.google.cloud.storage.JsonResumableSession.put(JsonResumableSession.java:68)
	at com.google.cloud.storage.ApiaryUnbufferedWritableByteChannel.internalWrite(ApiaryUnbufferedWritableByteChannel.java:115)
	at com.google.cloud.storage.ApiaryUnbufferedWritableByteChannel.write(ApiaryUnbufferedWritableByteChannel.java:60)
	at com.google.cloud.storage.UnbufferedWritableByteChannelSession$UnbufferedWritableByteChannel.write(UnbufferedWritableByteChannelSession.java:31)
	at com.google.cloud.storage.DefaultBufferedWritableByteChannel.write(DefaultBufferedWritableByteChannel.java:141)
	at com.google.cloud.storage.StorageByteChannels$SynchronizedBufferedWritableByteChannel.write(StorageByteChannels.java:131)
	at com.google.cloud.storage.BaseStorageWriteChannel.write(BaseStorageWriteChannel.java:119)
	... 6 more
Caused by: com.google.api.client.http.HttpResponseException: 404
PUT http://localhost:65555/b/my-staging-bucket/o/fake_file_name.zip
	at com.google.cloud.storage.JsonResumableSessionPutTask.call(JsonResumableSessionPutTask.java:194)
```


**Analysis**
This error is caused by `FakeStorageRpc l. 631`. The regex-pattern created in l. 629 should match the contentRange header. But in case the content does exceed the buffer-size the contentRange header is of type HttpContentRange.Incomplete. A `getHeaderValue()` call to a contentRange header of this type will return some String in this format `"bytes %d-%d/*"` (s. stacktrace where this equals `"bytes 0-262143/*"`)
This does not match with the regex pattern and therefore results in a `NOT_FOUND` response code being set in `FakeStorageRpc l. 644`.


**Proposed solution**
Use a corrected regex like `"bytes (?:\\d+-\\d+|\\*)/(\\d+|\\*)"`


**Dependency setup**
* `com.google.cloud:google-cloud-core:2.49.0`
* `com.google.cloud:google-cloud-storage:2.46.0`
* `com.google.cloud:google-cloud-nio:0.127.28`

--- a/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/testing/FakeStorageRpc.java
+++ b/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/testing/FakeStorageRpc.java
@@ -355,6 +355,7 @@ class FakeStorageRpc extends StorageRpcTestBase {
     }
     checkGeneration(key, generationMatch);
     metadata.put(key, object);
+    futureContents.put(key, new byte[0]);
 
     return key;
   }
@@ -615,7 +616,7 @@ class FakeStorageRpc extends StorageRpcTestBase {
           this.getStreamingContent().writeTo(baos);
           bytes = futureContents.get(key);
           byte[] byteArray = baos.toByteArray();
-          if (storageObject.getGeneration() != null) {
+          if (bytes != null) {
             futureContents.put(key, concat(bytes, byteArray));
           } else {
             futureContents.put(key, byteArray);
@@ -624,21 +625,27 @@ class FakeStorageRpc extends StorageRpcTestBase {
           List<String> contentRanges = getHeaders().get("content-range");
           if (contentRanges != null && !contentRanges.isEmpty()) {
             String contentRange = contentRanges.get(0);
-            // finalization PUT
-            // Pattern matches both froms of begin-end/size, */size
-            Pattern pattern = Pattern.compile("bytes (?:\\d+-\\d+|\\*)/(\\d+)");
-            Matcher matcher = pattern.matcher(contentRange);
-            if (matcher.matches()) {
-              storageObject.setSize(new BigInteger(matcher.group(1), 10));
+            if ("bytes */*".equals(contentRange) || contentRange.endsWith("/*")) {
+              // query or incremental put
+              resp.addHeader(
+                  "Range", String.format("bytes=0-%d", futureContents.get(key).length - 1));
+              resp.setStatusCode(308);
+            } else if (contentRange.startsWith("bytes */")
+                || contentRange.matches("bytes \\d+-\\d+/\\d+$")) {
+              // finalize
+              byte[] finalBytes = futureContents.get(key);
+              BigInteger size = BigInteger.valueOf(finalBytes.length);
               storageObject.setGeneration(System.currentTimeMillis());
               DateTime now = now();
               storageObject.setTimeCreated(now);
               storageObject.setUpdated(now);
+              storageObject.setSize(size);
               String string = GsonFactory.getDefaultInstance().toPrettyString(storageObject);
               resp.addHeader("Content-Type", "application/json;charset=utf-8");
               resp.addHeader("Content-Length", String.valueOf(string.length()));
               resp.setContent(string);
-              contents.put(key, byteArray);
+              resp.setStatusCode(200);
+              contents.put(key, finalBytes);
               futureContents.remove(key);
             } else {
               resp.setStatusCode(NOT_FOUND);

--- a/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/testing/FakeStorageRpcTest.java
+++ b/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/testing/FakeStorageRpcTest.java
@@ -21,6 +21,7 @@ import com.google.cloud.WriteChannel;
 import com.google.cloud.storage.Blob;
 import com.google.cloud.storage.BlobId;
 import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.HttpStorageOptions;
 import com.google.cloud.storage.Storage;
 import com.google.cloud.storage.Storage.BlobWriteOption;
 import com.google.cloud.storage.StorageOptions;
@@ -28,6 +29,7 @@ import com.google.cloud.storage.contrib.nio.testing.LocalStorageHelper.FakeStora
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
 import org.junit.Test;
 
 public final class FakeStorageRpcTest {
@@ -64,5 +66,26 @@ public final class FakeStorageRpcTest {
     String gen2read1 = new String(gen2.getContent(), StandardCharsets.UTF_8);
     System.out.println("gen2read1 = " + gen2read1);
     assertThat(gen2read1).isEqualTo("def");
+  }
+
+  @Test
+  public void multiChunkUploadWorks() throws Exception {
+    FakeStorageRpcFactory serviceRpcFactory = new FakeStorageRpcFactory();
+    HttpStorageOptions options =
+        StorageOptions.http().setServiceRpcFactory(serviceRpcFactory).build();
+    try (Storage storage = options.getService()) {
+
+      byte[] bytes = new byte[256 * 1024 + 37];
+      Arrays.fill(bytes, (byte) 'A');
+
+      BlobId id = BlobId.of("bucket", "object");
+      BlobInfo info = BlobInfo.newBuilder(id).build();
+      try (WriteChannel writeChannel = storage.writer(info)) {
+        writeChannel.setChunkSize(256 * 1024);
+        writeChannel.write(ByteBuffer.wrap(bytes));
+      }
+      byte[] actual = storage.readAllBytes(id);
+      assertThat(actual).isEqualTo(bytes);
+    }
   }
 }

